src/diffusers/pipelines/versatile_diffusion/pipeline_versatile_diffusion.py

```diff
@@ -0,0 +1,312 @@
+# Copyright 2022 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import inspect
+from typing import List, Optional, Tuple, Union
+
+import numpy as np
+import torch
+import torch.utils.checkpoint
+
+import PIL
+from transformers import CLIPProcessor, CLIPTextModel, CLIPTokenizer, CLIPVisionModel
+
+from ...models import AutoencoderKL, UNet2DConditionModel, VQModel
+from ...models.attention import Transformer2DModel
+from ...pipeline_utils import DiffusionPipeline, ImagePipelineOutput
+from ...schedulers import DDIMScheduler, LMSDiscreteScheduler, PNDMScheduler
+
+
+class VersatileMixedModel:
+    """
+    A context managet that swaps the transformer modules between the image and text unet during inference, depending on
+    the latent type and condition type.
+    """
+
+    def __init__(self, image_unet, text_unet, latent_type, condition_type):
+        self.image_unet = image_unet
+        self.text_unet = text_unet
+        self.latent_type = latent_type
+        self.condition_type = condition_type
+
+    def swap_transformer_modules(self):
+        for name, module in self.image_unet.named_modules():
+            if isinstance(module, Transformer2DModel):
+                parent_name, index = name.rsplit(".", 1)
+                index = int(index)
+                self.image_unet.get_submodule(parent_name)[index], self.text_unet.get_submodule(parent_name)[index] = (
+                    self.text_unet.get_submodule(parent_name)[index],
+                    self.image_unet.get_submodule(parent_name)[index],
+                )
```

Comment:
This follows the original codebase: the attention modules are swapped between the text and image unets at inference time, depending on what we generate (image or text) and what we use for conditioning (CLIP image or text embeddings).
One difference: instead of [enumerating all of the resnet modules during the forward pass](https://github.com/SHI-Labs/Versatile-Diffusion/blob/853d555d282180e911da69e276d40a68a46da770/lib/model_zoo/openaimodel.py#L2506) we just swap the transformer weights.

Reply:
Are the weights for the respective unets identical?

Reply:
The weights (resents, transformers, etc) are fully different between the unets. However the 2D transformer blocks are swapped depending on the task (both at train and inference time)